### PR TITLE
[ownership] Loosen restrictions around what we specialize and add generic specialization tests behind a flag.

### PR DESCRIPTION
**Pull request by gottesmm**
The idea is that this will let me remove these assertions that were in place to
make sure we were really conservative around specializing ownership code. For me
to remove that I need to be able to actually test out this code (since I think
there are some code paths where this will trigger in other parts of the compiler
now).

So to work out the kinks, I added a flag that allows for the generic specializer
to process ownership code and translated most of the .sil test cases/fixed any
bugs that I found. This hopefully will expose anything that is missing.

NOTE: I have not enabled the generic specializer running in ownership in the
pipeline. This is just a step in that direction by adding tests/etc.

### CODE REVIEW

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 1]
```diff
@@ -1937,6 +1942,15 @@ static void prepareCallArguments(ApplySite AI, SILBuilder &Builder,
   }
 }
 
+static void cleanupCallArguments(SILBuilder &builder, SILLocation loc,
+                                 ArrayRef<SILValue> values) {
+  for (SILValue v : values) {
+    if (auto *lbi = dyn_cast<LoadBorrowInst>(v)) {
+      builder.createEndBorrow(loc, lbi);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Generics.cpp:1949]: I am a little nervous about this. I think I am going to change values here to have a bit that says it needs an end borrow or something like that.

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 2]
```diff
@@ -2392,11 +2426,24 @@ void swift::trySpecializeApplyOfGeneric(
...
-    for (Operand *Use : PAI->getUses()) {
+    SmallVector<Operand *, 4> worklist(PAI->getUses());
+    while (!worklist.empty()) {
+      auto *Use = worklist.pop_back_val();
+
       SILInstruction *User = Use->getUser();
+
+      // Look through copy_value.
+      if (auto *cvi = dyn_cast<CopyValueInst>(User)) {
+        llvm::copy(cvi->getUses(), std::back_inserter(worklist));
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Generics.cpp:2437]: I added this b/c some of the tests around recursive specialization started failing b/c it couldn't look through these instructions/ignore incidental instructions.

File: lib/SILOptimizer/Utils/Generics.cpp (hunk 1 above)
Comment (eeckstein) [reply lib/SILOptimizer/Utils/Generics.cpp:1949]: You are right: what if there was a load_borrow before specialization? The you would insert a wrong end_borrow

**Review by eeckstein — CHANGES_REQUESTED** [review]
basically lgtm.
Request for changes only for a better way to decide on inserting an end_borrow (see my comment)

**Comment by gottesmm** [comment]
@eeckstein I fixed it by using a SmallBitVector. I learned about something new that I didn't know about: SmallBitVector now has an iterator for set bit indices! So you can just do:

```
for (int index : bitVector.set_bits()) { ... }
```

instead of:

```
for (int index = bitVector.find_first(); index != -1; index = bitVector.find_next(index)) { ... }
```

**Comment by gottesmm** [comment]
@swift-ci test

File: lib/SILOptimizer/Utils/Generics.cpp [hunk 3]
```diff
@@ -1969,6 +1997,10 @@ static ApplySite replaceWithSpecializedCallee(ApplySite applySite,
     assert(resultBlock->getSinglePredecessorBlock() == tai->getParent());
     auto *newTAI = builder.createTryApply(loc, callee, subs, arguments,
                                           resultBlock, tai->getErrorBB());
+    FullApplySite(newTAI).insertAfterFullEvaluation(
+        [&](SILBasicBlock::iterator insertPt) {
+          cleanupCallArguments(builder, loc, arguments, argsNeedEndBorrow);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/Generics.cpp:2002]: Hmmm... I think I need to create a new SILBuilder here.

Comment (gottesmm) [reply lib/SILOptimizer/Utils/Generics.cpp:2002]: I am going to try to break this. This to not fail says we are missing code coverage.

**Review by eeckstein — APPROVED** [review]
lgtm!

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
I was correct that we were missing code coverage. I added the missing code coverage of try_apply/begin_apply.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test windows platform

**Comment by gottesmm** [comment]
@swift-ci test OS X platform

**Comment by gottesmm** [comment]
OS X failure was a sandbox issue

**Comment by gottesmm** [comment]
I took a look at why we are failing/hitting the verifier error. The reason why is that there is a debug_value on self that is after the store [init] of self at the beginning of a function to materialize the argument.

**Comment by gottesmm** [comment]
Tracked down the problem. The issue was that I was creating a debug_value in the cloner too late. I put in a fix and also put in some things to make it easier to debug this sort of thing in the future:

1. A constructor for PrettyStackTraceSILFunction that takes a llvm::Twine so we can make more complex messages.
2. Added an -Xllvm flag that causes us to verify when we produce specializations and also added PrettyStackTraceSILFunction for the original unspecialized SIL function so if we hit a crash we can see the input to the specializer. That turned out to be very useful for me.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

File: lib/SILOptimizer/Utils/GenericCloner.cpp [hunk 4]
```diff
@@ -106,24 +107,36 @@ void GenericCloner::populateCloned() {
...
+          // store below.
           for (Operand *ArgUse : OrigArg->getUses()) {
             if (auto *DVAI = dyn_cast<DebugValueAddrInst>(ArgUse->getUser())) {
+              auto *oldScope = getBuilder().getCurrentDebugScope();
               getBuilder().setCurrentDebugScope(
                   remapScope(DVAI->getDebugScope()));
               getBuilder().createDebugValue(DVAI->getLoc(), NewArg,
                                             *DVAI->getVarInfo());
-              getBuilder().setCurrentDebugScope(nullptr);
+              getBuilder().setCurrentDebugScope(oldScope);
```
Comment (gottesmm) [inline lib/SILOptimizer/Utils/GenericCloner.cpp:126]: @adrian-prantl @dcci @vedantk did I do this right? I don't understand how setting this to nullptr made sense before.

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test windows platform

**Comment by gottesmm** [comment]
@swift-ci test windows platform

File: lib/SILOptimizer/Utils/GenericCloner.cpp (hunk 4 above)
Comment (vedantk) [reply lib/SILOptimizer/Utils/GenericCloner.cpp:126]: It looks like GenericCloner::postProcess fixes up the scopes after populateCloned is finished. SILCloner isn't equipped to handle debug insts without scopes, hence the special-casing here. I'd expect you to be able to delete the second setCurrentDebugScope call.

I'm not sure why the cloning isn't structured in a way that allows remapping scopes as instructions get copied over.

**Comment by gottesmm** [comment]
Reproduced the error.

**Comment by gottesmm** [comment]
Interesting:

```
// thunk for @escaping @callee_guaranteed (@inout A, @guaranteed PathArgument) -> (@error @owned Error)
sil shared [transparent] [serializable] [reabstraction_thunk] [ossa] @$sx10TSCUtility12PathArgumentVs5Error_pIeglgzo_xACsAD_pIeglnzo_8Commands11ToolOptionsCRbzlTR : $@convention(thin) <Options where Options : ToolOptions> (@inout Options, @in_guaranteed PathArgument, @guaranteed @callee_guaranteed (@inout Options, @guaranteed PathArgument) -> @error Error) -> @error Error {
// %0                                             // user: %4
// %1                                             // user: %3
// %2                                             // user: %4
bb0(%0 : $*Options, %1 : $*PathArgument, %2 : @guaranteed $@callee_guaranteed (@inout Options, @guaranteed PathArgument) -> @error Error):
  %3 = load_borrow %1 : $*PathArgument            // users: %7, %10, %4
  try_apply %2(%0, %3) : $@callee_guaranteed (@inout Options, @guaranteed PathArgument) -> @error Error, normal bb1, error bb2 // id: %4

bb1(%5 : $()):                                    // Preds: bb0
  %6 = tuple ()                                   // user: %8
  end_borrow %3 : $PathArgument                   // id: %7
  return %6 : $()                                 // id: %8

// %9                                             // user: %11
bb2(%9 : @owned $Error):                          // Preds: bb0
  end_borrow %3 : $PathArgument                   // id: %10
  throw %9 : $Error                               // id: %11
} // end sil function '$sx10TSCUtility12PathArgumentVs5Error_pIeglgzo_xACsAD_pIeglnzo_8Commands11ToolOptionsCRbzlTR'
```

**Comment by gottesmm** [comment]
Bug in the reabstract thunk generation code.

**Comment by gottesmm** [comment]
Found the bug. It was my bug. Thank god for ownership verification = ).

**Comment by gottesmm** [comment]
I am going to try again tonight.

**Comment by gottesmm** [comment]
Tracked down the problem. The problem was that there was some misleading code that seemed like it was only handling apply inst. Instead it was also handling try_apply, but returning the normal block argument as the "result". This seems like a general useful property, so I put it into a helper on ApplySite. Now it all works.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test windows platform

**Comment by gottesmm** [comment]
Found more missing test coverage. We don't ever generate one of these reabstraction thunks with `@out` parameters in the test suite...

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test source compatibility

**Comment by gottesmm** [comment]
@swift-ci test windows platform

**Comment by gottesmm** [comment]
@swift-ci test windows platform

**Comment by gottesmm** [comment]
Looks like Source Compatibility matches the state on ci.swift.org. It isn't this commit.

**Pull request merged**